packages/shared/scope-hoisting/src/generate.js

```diff
@@ -73,21 +90,33 @@ export function generate({
   // at the right time (after other bundle dependencies are loaded).
   let statements = ast.program.body;
   if (bundle.env.outputFormat === 'global') {
-    statements = isAsync
-      ? [
-          REGISTER_TEMPLATE({
-            STATEMENTS: statements,
-            REFERENCED_IDS: t.arrayExpression(
-              [mainEntry, ...referencedAssets]
-                .filter(Boolean)
-                .map(asset =>
-                  t.stringLiteral(bundleGraph.getAssetPublicId(asset)),
-                ),
-            ),
-            PARCEL_REQUIRE: t.identifier(parcelRequireName),
-          }),
-        ]
-      : [WRAPPER_TEMPLATE({STATEMENTS: statements})];
+    // Wrap async bundles in a closure and register with parcelRequire so they are executed
+    // at the right time (after other bundle dependencies are loaded).
+    if (isAsync) {
+      statements = REGISTER_TEMPLATE({
+        STATEMENTS: statements,
+        REFERENCED_IDS: t.arrayExpression(
+          [mainEntry, ...referencedAssets]
+            .filter(Boolean)
+            .map(asset => t.stringLiteral(bundleGraph.getAssetPublicId(asset))),
+        ),
+        PARCEL_REQUIRE: t.identifier(parcelRequireName),
+      });
+    }
+
+    if (needsPrelude(bundle, bundleGraph)) {
```

Comment:
Maybe easier to add this in `link` and avoid `hoistedCalls` being passed around?